Add "Display layout editor"

LunaMoo opened the pull request:
To use on tablets, or when people want to use integer zoom while keeping black bars or cropping image. By default it's set to do an automatic cropping when stretching is disabled since it replaces the 1080p exception.

 It's so big because I had to rebuilt atlas to add tiny half kb image.:C Not sure if I rebuilt atlas correctly since there were a few missing signs and I got arialuni.ttf from the net since I didn't had it.

 Everything works, althrough I'm unsure of the options names and probably it might require some changes. I mostly based the editor screen on touch control, so there was some copy&paste involved, althrough I stripped it out of what this feature didn't need.
It looks like that:
![edit screen](https://cloud.githubusercontent.com/assets/5485237/10561991/108220a6-7545-11e5-8687-30fcebf9b262.jpg)
It's also aware of the display rotation, also the image that symbolizes the display is transparent and the lines symbolize screen bounds, which can be passed for cropping image.

 I guess it will collide with combo key pull request since it also rebuilds atlas, but as I said I'm unsure if my arialuni font was correct version, so would be better if someone takes this over.:3 Any changes are welcome, I didn't really wrote that for myself and I'm undecided how it should look like.

Edit: Improved a bit of stupidity/simplified the code I made while being half asleep ~_~(heh side effect from training my body too much, I don't really feel when my brain slows down:C).
 Also changed boundaries to avoid moving the game display out of the view completely as someone could do that accidently and then wonder why he sees black screen. Moving it half way every direction should be more than enough even for somewhat extreme image crop/zoom in.

hrydgard commented:
Nice! I'll take this over later and rebuild the atlas after fixing and merging the combo keys one.

unknownbrackets commented on GPU/Common/FramebufferCommon.cpp:
```diff
@@ -40,16 +40,44 @@ void CenterRect(float *x, float *y, float *w, float *h, float origW, float origH
-			*h = 1088;
-			return;
+		bool fullScreenZoom = true;
+#ifdef _WIN32
```
Not Windows specific.  Maybe MOBILE_DEVICE or something?

I see now this matches here, which is possibly also wrong?
https://github.com/hrydgard/ppsspp/blob/953f712e3a8efd615b1b40a1f25179fdf8a5d937/Core/Config.cpp#L438

SDL / Qt are not always full screen either (Mac, Linux.)

Also, I think we should rename `CenterRect` to like `CenterDisplayOutputRect` or something, to make sure it's never used for another purpose.

-[Unknown]

LunaMoo replied:
Oh yeah I get the fullscreen idea by looking at the config. Will change the config one as well then.

 CenterRect was always like that, I can just change it everywhere I guess, I just wonder about this https://github.com/hrydgard/ppsspp/blob/c2f4fad56aada331bbd8dd7db99037285bab853f/GPU/Software/SoftGpu.cpp#L52 is this code even used? Because that feature I added seems to work also in software mode.

unknownbrackets replied:
It's used here, but should be from Common:
https://github.com/hrydgard/ppsspp/blob/c2f4fad56aada331bbd8dd7db99037285bab853f/GPU/Software/SoftGpu.cpp#L250

We can probably remove that definition.

-[Unknown]

LunaMoo replied:
Ok:3, done that as well.

unknownbrackets commented on GPU/Common/FramebufferCommon.cpp:
```diff
@@ -40,16 +40,44 @@ void CenterRect(float *x, float *y, float *w, float *h, float origW, float origH
+			if (g_Config.iSmallDisplayZoom != 1) {
+				float offsetX = (g_Config.fSmallDisplayOffsetX - 0.5f) * 2.0f * frameW;
+				float offsetY = (g_Config.fSmallDisplayOffsetY - 0.5f) * 2.0f * frameH;
+				float customZoom = g_Config.fSmallDisplayCustomZoom / 8.0f;
```
Why is this using an 8.0f multiple everywhere?

-[Unknown]

LunaMoo replied:
Because I use it also for image scale which is smallest possible while keeping psp proportions:)

unknownbrackets commented on Core/Config.h:
```diff
@@ -144,7 +144,10 @@ struct Config {
+	int iSmallDisplayZoom;  // Used to fit display into screen 0 = disabled, 1 = auto, 2 = x1, 3 = x2 etc..
+	float fSmallDisplayOffsetX;
+	float fSmallDisplayOffsetY;
+	float fSmallDisplayCustomZoom;
```
Do these only make the display smaller?  In the case of 1080p, it's technically making it larger, right?  Theoretically the zoom can go either way?

-[Unknown]

LunaMoo replied:
Yeah, it can do zoom in and zoom out, I could word it differently maybe, fSmallDisplayCustomZoom is actually used for the zoom, so fractions are possible as well, iSmallDisplayZoom is only used inside the editor for integer zoom which also turns on manual edit.

unknownbrackets commented on GPU/Common/FramebufferCommon.cpp:
```diff
@@ -40,16 +40,44 @@ void CenterRect(float *x, float *y, float *w, float *h, float origW, float origH
+				}
+			} else {
+				if (!rotated) {
+					float pixelCrop = frameH / 270.0f;
```
For clean multiples of HD, sure, it works like that (although technically you want the remainder imho, the math works out either way), but consider someone with an unclean multiple - like 2560x1600.  It's not like 4:3 and other ratios no longer exist.

At 1600, this would be about 5.93.  That would give some pixel offset blur, and the height would be 1611.85 or so, which would stretch the display vertically, I suppose?

-[Unknown]

LunaMoo replied:
The code under this change was dealing with those, I guess I should just implement that code inside my change and maybe remove the option to disable it since it was there only for that.

LunaMoo replied:
I will just do it like:

    				float resCommonWidescreen = pixelCrop - round(pixelCrop);
    				if (!rotated && resCommonWidescreen == 0) {

and remove "disable" option for this, so 0 will be auto, any other case that doesn't fit common widescreen will use either earlier code that's under this or whatever was manually selected zoom:) Should be fine.

LunaMoo commented:
Forgot in commit changelog, but also changed fullscreen in config as suggested by [Unknown] to #ifndef MOBILE_DEVICE
:3

hrydgard commented:
@LunaMoo  You can rebase on master and get rid of your atlas changes - I threw in psp_display.png in the atlas over in the Combo Key pull, which I've now merged.

LunaMoo committed:
Add Display layout editor to use for example on tablets, but also includes automatic screen cropping for 16:9 cases(default).
//Rebased & removed atlas stuff which was added with #8096
//squashed with some improvements

LunaMoo commented:
Done:3, thanks for adding the png earlier:).

hrydgard commented:
@LunaMoo looks like you forgot to add DisplayLayoutEditor and DisplayLayoutScreen.cpp? :)

LunaMoo commented:
Yeah sorry^^.

hrydgard commented:
Hm, there seems to be some bugs with DPI:  Auto mode does not preview correctly if DPI is not 1.0 (like on most Android devices). It does seem to work correctly in-game though.

LunaMoo commented:
I think I only made the auto stretch for looks and probably set it only after it's selected, will check in a moment.

xiushudongfang commented:
Nice,very useful function.  There are some bugs with skipping buffer effects.

Buffered rendering:
![screenshot_2015-10-24-21-25-04](https://cloud.githubusercontent.com/assets/8541580/10710686/ef2bb9f6-7a96-11e5-9fa0-6343c56eef6d.png)
skipping buffer effects:
![screenshot_2015-10-24-21-25-16](https://cloud.githubusercontent.com/assets/8541580/10710697/3e96800c-7a97-11e5-8e4d-a44167b58e47.png)

LunaMoo commented:
Initialized auto mode preview, not sure yet why it cuts image in non-buffered rendering:|.

LunaMoo commented:
Hopefully the way I fixed this is acceptable, turns out #7629 broke "small display" in non buffered rendering ages ago, suprising nobody reported that.

unknownbrackets commented:
Hmm, I see.  Yeah, I guess that makes sense.

-[Unknown]

xiushudongfang commented:
There is another bug in non buffered rendering.
The layout screen :
![screenshot_2015-10-25-00-40-05](https://cloud.githubusercontent.com/assets/8541580/10711611/cee0ee88-7ab3-11e5-8449-e55acc84e6ac.png)
The game screen :
![screenshot_2015-10-25-00-40-45](https://cloud.githubusercontent.com/assets/8541580/10711616/e9df0a76-7ab3-11e5-89f7-c44848391bee.png)

You can see that the position of the screen is wrong.The Y axis is inverse.

LunaMoo commented:
Hmmm this is on windows, android or what? I saw it inverted on windows 7 here, so I corrected that already, maybe it's different for GLES?

xiushudongfang commented:
Android

LunaMoo commented:
That should deal with it, hopefully it's same for all mobile devices.

xiushudongfang commented:
Thanks ！

LunaMoo commented:
You're welcome, hopefully everything works now.

hrydgard commented:
That is mysterious. The GL coordinate system should be the same...

LunaMoo commented:
Would be nice if someone could test it with non buffered rendering on any other OGL platform like linux, just in case that define should only be for windows vs everything else.
 I couldn't really solve that mystery, but I'm thinking maybe it's just some confusion on windows, because we have 2 backends on it and it's easy to forget about one or another. D3D doesn't have any coordinate difference between buffered and non buffered modes.

xiushudongfang commented:
The Y axis  inverse bug may not be fixed on Android in  non buffered rendering. It works well only in CG.

LunaMoo commented:
Thanks for testing, I checked FMV's now and current master(v1.1.1-94-g6b905e8) is actually broken with non buffered rendering on windows:
 - FMV's are broken in D3D non buffered rendering ~ they look zoomed in,
 - they are also broken in OGL with non buffered rendering ~ displaying two copies, the second one is 480x272 sticked to bottom left corner.

 With this branch on windows they're broken same way + inverted coords. Heh, what a messup. :C

unknownbrackets commented:
Those will likely be copied from a source that isn't flipped.

Here, flipping isn't needed, since the source is a right-side-up texture:
https://github.com/LunaMoo/ppsspp/blob/DisplayEditing/GPU/GLES/Framebuffer.cpp#L407

Here we are displaying an inverted texture (FBO / render target):
https://github.com/LunaMoo/ppsspp/blob/DisplayEditing/GPU/GLES/Framebuffer.cpp#L1038

So you might need a flag from the caller to get it right.

-[Unknown]

LunaMoo commented:
Thanks for the pro tip, added another parameter to CenterDisplayOutputRect so I can invert each and every call separately.
Pretty much when I flip the non buffered rendering I have to avoid flipping this:
 https://github.com/LunaMoo/ppsspp/blob/DisplayEditing/GPU/GLES/Framebuffer.cpp#L407
and it works for fmv's, assuming same thing for android I guess if currently only FMV's were placed correctly, so it will also simplify the define to windows and everything else(since on windows I don't want to do that with D3D).

Still FMV's are broken with current master which is not affected by this branch, not sure how to fix those issues. I'll see if I can find what broke them after I figure this out, never really use non buffered rendering so I don't know if it's new regression or not.:]

LunaMoo commented:
Hmmm the issues with FMV's I had were partially modified by me using same config with older versions I tested, with fresh config the small FMV window from bottom left, covers whole screen now so scaling and resizing appears to not work at all in them with non buffered rendering, have to investigate further.:]

LunaMoo commented:
![double fmv](https://cloud.githubusercontent.com/assets/5485237/10714503/4a27d474-7af2-11e5-9e23-79f69ef44874.jpg)
Here's an image from official build, when using x1 rendering resolution and small display feature in non buffered rendering. ~_~
1 - is affected by small display(and in my branch by all the scalling and position),
2 - is affected by rendering resolution - always sits in bottom left corner(upper left in D3D).

I belive it was always done like that saw it by testing some 0.9.8+ as well, really weird thing, not sure how to properly remove the 2nd one, any hints are welcome.

hrydgard commented:
Hm, looks like we have some bugs indeed :) Maybe we should fix it in master before trying to merge this. I will take a look soon.

LunaMoo commented:
Thanks, my ideas for this now would at best lead to some hiding hackery, need to learn more.;p

hrydgard commented:
@LunaMoo , I think I've got all the small screen bugs in master now. Please rebase :)

LunaMoo commented:
Thanks, rebased now, also removed the non buffered rendering workaround in state mapping since it wasn't needed anymore with your fixes:).

 I noticed the issue mentioned earlier about double image in FMV's + non buffered rendering still does happen, but that's just my weird luck - last time I only tested God Eater 2 and apparently from testing today that issue happens only in that game, or at least I could only see it in that one among games I tested. Will check if it can be reproduced in demo since that game is broken anyway and might just be doing something weird.

hrydgard commented:
@LunaMoo , fixed that for you :)

LunaMoo commented:
Cool, thanks, will rebase again then:3.

LunaMoo committed:
Add video fix in non buffered mode for D3D backend as well.

LunaMoo commented:
Done, added your fix to D3D backend as well, with that, all modes, hw and sw works fine:). I guess it's ready for merging now.

unknownbrackets commented on UI/GameSettingsScreen.cpp:
```diff
@@ -774,6 +779,11 @@ UI::EventReturn GameSettingsScreen::OnFullscreenChange(UI::EventParams &e) {
 }
 
+UI::EventReturn GameSettingsScreen::OnDisplayLayoutEditor(UI::EventParams &e) {
+	screenManager()->push(new DisplayLayoutScreen()); //TouchControlLayoutScreen());//
```
Can you remove this comment?  Sorry to nitpick.

-[Unknown]

unknownbrackets commented on UI/DisplayLayoutScreen.cpp:
```diff
@@ -0,0 +1,227 @@
+
+	mode_ = new ChoiceStrip(ORIENT_VERTICAL, new AnchorLayoutParams(leftColumnWidth, WRAP_CONTENT, 10, NONE, NONE, 158 + 64 + 10));
+	if (g_Config.iSmallDisplayZoom == 0) {
+		mode_->AddChoice(gr->T("Activated(Auto)"));
```
Can we make this "Active (Auto)" or something (with the space)?  What do you think?

-[Unknown]

LunaMoo committed:
Small edits, nothing in the code

LunaMoo commented:
@unknownbrackets sorry for the comment, part of copy paste I forgot about:), and yeah any changes how it looks are welcome. I made it initially for just some sliders in graphics menu and when Henrik suggested in the forum to make a touch screen for it, I kind of went for it without any plan:).

hrydgard commented:
Hm, I just tested this again, and it doesn't seem to work anymore? It may have to wait..

hrydgard commented:
Ah, it only works in full screen. Not sure I like that - I think I want it to apply when windowed too..

LunaMoo commented:
:3 Just wanted to ask if you tested in full screen when it updated with merge.

 You can change it if you want or I can open another pull request for it.

LunaMoo commented:
Coordinates got inverted after the flip got merged, also seems that GE2 double FMV returned?:| Will check everything and try to correct what I can.
